File: bcc-apps/python/bashreadline.c
/* Tracing bash's readline retval */
#include <uapi/linux/ptrace.h>

struct data_t {
	u32 uid;
	char command[64];
};
BPF_PERF_OUTPUT(events);

int bash_readline(struct pt_regs *ctx)
{
	struct data_t data = { };
	data.uid = bpf_get_current_uid_gid();

	// PT_REGS_RC(ctx) holds the return value.
	bpf_probe_read_user(&data.command, sizeof(data.command),
			    (void *)PT_REGS_RC(ctx));

	// submit perf event.
	events.perf_submit(ctx, &data, sizeof(data));
	return 0;
}

File: bcc-apps/python/bashreadline.py
#!/usr/bin/env python3
# Tracing bash's readline retval
from bcc import BPF
from time import strftime

# load BPF program
b = BPF(src_file="bashreadline.c")

# attach uretprobe
b.attach_uretprobe(name="/usr/bin/bash", sym="readline", fn_name="bash_readline")


# callback for perf event
def print_event(cpu, data, size):
    event = b["events"].event(data)
    print(
        "%-9s %-6d %s"
        % (strftime("%H:%M:%S"), event.uid, event.command.decode("utf-8"))
    )


# print header
print("%-9s %-6s %s" % ("TIME", "UID", "COMMAND"))

# loop with callback to print_event
b["events"].open_perf_buffer(print_event)
while 1:
    try:
        b.perf_buffer_poll()
    except KeyboardInterrupt:
        exit()
